### PR TITLE
Setup allreduce strategy toggle in Worker

### PR DESCRIPTION
**Pull request by terrytangyuan**
Signed-off-by: terrytangyuan <terrytangyuan@gmail.com>

### CODE REVIEW

File: docs/designs/allreduce.md [hunk 1]
```diff
@@ -65,15 +65,15 @@ The interface for allreduce would look like the following:
...
 
 for _ in range(num_epochs):
     for features, labels in data_loader:
         outputs = model(features)
         loss = loss_fn(outputs, labels)
         grads = calculate_gradients(loss, model)
-        res, averaged_grads = communicator.allreduce_average(grads)
-        if res == SUCCESS:
+        status, averaged_grads = communicator.allreduce_average(grads)
+        if res == SUCCEEDED:
```
Comment (skydoorkai) [inline docs/designs/allreduce.md]: res -> status

File: docs/designs/allreduce.md [hunk 2]
```diff
@@ -129,14 +129,14 @@ that we mentioned in the previous section. The main differences are the followin
...
+communicator = CollectiveCommunicator()
 
 with tf.GradientTape() as tape:
     outputs = model(features, training=True)
     loss = loss_fn(outputs, labels)
 local_grads = tape.gradient(loss, get_trainable_items())
-res, averaged_grads = communicator.allreduce_average(grads)
-if res == SUCCESS:
+status, averaged_grads = communicator.allreduce_average(grads)
+if res == SUCCEEDED:
```
Comment (skydoorkai) [inline docs/designs/allreduce.md]: res -> status

File: elasticdl/python/collective_ops/communicator.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# TODO: This is dummy for now until the real implementation
+# has been open sourced
+
+
+class CollectiveCommunicator(object):
+    def __init__(self):
+        pass
+
+    def allreduce_average(self, data):
```
Comment (skydoorkai) [inline elasticdl/python/collective_ops/communicator.py]: Better to use `allreduce(self, data, op)` where `op` can be `sum`, `average`, etc.

**Review by skydoorkai — COMMENTED** [review]
How to unittest the allreduce code path?

File: elasticdl/python/worker/worker.py [hunk 4]
```diff
@@ -575,12 +592,25 @@ def forward_process(self, features):
 
     def _run_training_task(self, features, labels):
         loss, grads = self.training_process(features, labels)
-        accepted, min_model_version = self.report_gradient(grads)
-        if accepted and self._get_model_steps > 1:
-            non_embed_vars_n = len(self._non_embed_vars)
-            self._non_embed_grads = grads[:non_embed_vars_n]
-        self._reset_embedding()
-        return accepted, min_model_version, loss
+        if self._distribution_strategy == DistributionStrategy.ALLREDUCE:
```
Comment (brightcoder01) [inline elasticdl/python/worker/worker.py:628]: Can we separate these two section of code into two methods?

**Comment by terrytangyuan** [comment]
@skydoorkai @brightcoder01 Thanks for the suggestions. I've addressed them and introduced additional tests to cover the new code.

File: elasticdl/python/collective_ops/communicator.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+
+class CollectiveCommunicator(object):
+    def __init__(self):
+        pass
+
+    def allreduce(self, data, op="MEAN"):
+        return CollectiveCommunicatorStatus.SUCCEEDED, data
+
+    def broadcast(self, from_worker_ip):
```
Comment (skydoorkai) [inline elasticdl/python/collective_ops/communicator.py:14]: How can the broadcast sender send some data using this method? `data` argument may also be needed.

**Review by skydoorkai — COMMENTED** [review]
Is the parameter initialization in workers in the next PR?

**Review by skydoorkai — APPROVED** [review]

**Pull request merged**